### DISCUSSION TITLE
On the possibility of updating ES's graphics system

### DISCUSSION THREAD

**Discussion by danaris** [endless-sky/endless-sky#10428] [Balance, Mechanics, and Interactions]
To re-post what I originally said in my issue report #10417:

> It's also worth noting that **OpenGL support is deprecated on macOS as a whole**. While _hopefully_ this particular bug can be resolved by something much simpler, I would be very interested in whether it's feasible to work on porting the graphics code of ES over to Vulkan, which is more modern and will continue to work on macOS through the MoltenVK library (which translates the Vulkan commands to Metal commands). The other potential option in this vein would be to abstract the graphics processing in such a way that we can have multiple rendering pipelines (selected at compile-time)—a Metal pipeline for macOS and the legacy OpenGL pipeline for other platforms. IMO, abstracting the graphics processing would be very valuable in any case. However, I recognize that any of these options would be a pretty monumental undertaking.

Though RisingLeaf is still having trouble with the bug, my success with it does _suggest_ that it may be due to misconfiguration rather than deeper issues for the moment. Thus, I felt it made sense to bring the broader issue of the graphics libraries being used to a Discussion, particularly since several contributors were at least open to the idea of such a refactor in the comments on the issue.

**Comment by RisingLeaf** [comment]
Some facts to the matter:
- The main options are openGL and Vulkan, while DirectX exists it is unsuited for our multi platform approach.

- The openGL code in ES is very spread out, a first step towards easing the switch or parallel support would be to centralise openGL commands in a meaningful manner (I will explain what meaningful means on request)

- vulkan uses pre compiled shaders, integrating the compilation step into cmake is easy but requires separate shader files 

- Vulkan doesn't use uniforms the way openGL does, while there are so called push constants they are very limited, it would make more sense to use uniform buffers which are basically used the same way by openGL and vulkan.

- vulkan gives more control than opengl, but this comes at the price of being way more verbose a single openGL line might equal twenty or more in vulkan

- vulkan can be multithreaded, which is cool 😎 

Summary: I don't dislike the idea of switching or having two render backends but it is a huge process and a plan should be made before hand, including the design of the render backends and how we want our pipeline to look.
E.G. uniform buffers can be accessed from multiple shaders, do we have universal variables that could be used by all shaders?
